rgw: Add rgw rate limiting per user and per bucket (#42891)
ceph/ceph: doc/radosgw/admin.rst

   be restarted for the changes to take effect.

> Rate Limit Management

Reviewer: are these configuration per tenant or across tenants?

Author: per tenant

Reviewer: so please add this information in the doc

=====================

The Ceph Object Gateway enables you to set rate limits on users and buckets. 

> rate limit include the maximum number of read ops and write ops per minute

Reviewer: ```suggestion
rate limit includes the maximum number of read ops and write ops per minute
```

> and how many write bytes or read bytes per minute a user or bucket can make.

Reviewer: ```suggestion
and how many bytes per minute could be written or read per user or bucket.
```

> Requests that are using GET or HEAD method in the REST request considered as read, otherwise it is write.

Reviewer: ```suggestion
Requests that are using GET or HEAD method in the REST request are considered as "read requests", otherwise they are considered as "write requests".
```

> Every Object Gateway will track after the user/bucket usage itself and will not share those metrics with other gateways.

Reviewer: ```suggestion
Every Object Gateway tracks per user/bucket separately, and these metrics are not shared metrics other gateways.
Also, note that usage is calculated after an operation was successfully finished.
```

That means that the desired limits configured should be divide by the number of active Object Gateways.

> if userA should be with a limit 10 ops per minute and there are 2 Object Gateways in the cluster,

Reviewer: ```suggestion
For example, if userA should be limited by 10 ops per minute and there are 2 Object Gateways in the cluster,
```

the limit over userA should be 5 (10 ops per minute / 2 RGWs).

> 

Reviewer: please add a note here, explaining that if the requests are not balanced between RGWs, the rate limit may be underutilized.

> - **Bucket:** The ``--bucket`` option allows you to specify a rate limit for

Reviewer: is this mandatory when the scope is "bucket"?

Author: yes , but I don't think it is worth mentioning because it's on the bucket rate limit set

  bucket.

> - **Maximum Read Ops:** The ``--max-read-ops`` setting allows you to specify

Reviewer: is it mandatory to set at least one of the "max" values?

Author: It is mandatory, changed the cli examples to reflect it

  the maximum number of read ops per minute per RGW. A 0 value disables this setting (which means unlimited access).
  
- **Maximum Read Bytes:** The ``--max-read-bytes`` setting allows you to specify
  the maximum number of read bytes per minute per RGW. A 0 value disables this setting (which means unlimited access).

- **Maximum Write Ops:** The ``--max-write-ops`` setting allows you to specify
  the maximum number of write ops per minute per RGW. A 0 value disables this setting (which means unlimited access).
  
- **Maximum Write Bytes:** The ``--max-write-bytes`` setting allows you to specify
  the maximum number of write bytes per minute per RGW. A 0 value disables this setting (which means unlimited access).
 
- **Rate Limit Scope:** The ``--ratelimit-scope `` option sets the scope for the rate limit.
  The options are ``bucket`` , ``user`` and ``anonymous``. Bucket rate limit apply to buckets. 

>   User quotas apply to a user. Anonymous apply to an unauthenticated user.

Reviewer: ```suggestion
  The user rate limit applies to a user. Anonymous apply to an unauthenticated user.
```

  Anonymous scope is only available for global rate limit.

> 

Reviewer: what about `--uid` ?

Set User Rate Limit
-------------------

Before you enable a rate limit, you must first set the rate limit parameters.
For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=user --uid=<uid> [--max-read-ops=<num ops>] [--max-read-bytes=<num bytes>]
  [--max-write-ops=<num ops>] [--max-write-bytes=<num bytes>]

For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=user --uid=johndoe --max-read-ops=1024 --max-write-bytes=10240

A 0 value for num ops and / or num bytes means that the
specific rate limit attribute check is disabled.

Get User Rate Limit
-------------------

Get the current configured rate limit parameters
For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=user --uid=<uid>

For example:: 

	radosgw-admin ratelimit get --ratelimit-scope=user --uid=johndoe

A 0 value for num ops and / or num bytes means that the
specific rate limit attribute check is disabled.

Enable/Disable User Rate Limit
------------------------------

> Once you set a user rate limit, you may enable it. For example:: 

Reviewer: what if you set the global rate limits and then enable the user? will the global settings apply here?

	radosgw-admin ratelimit enable --ratelimit-scope=user --uid=<uid>

You may disable an enabled user rate limit. For example:: 

	radosgw-admin ratelimit disable --ratelimit-scope=user --uid=johndoe

Set Bucket Rate Limit
---------------------

> Before you enable a rate limit, you must first set the rate limit parameters.

Reviewer: same as https://github.com/ceph/ceph/pull/42891/files#r771946110

For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=bucket --bucket=<bucket> [--max-read-ops=<num ops>] [--max-read-bytes=<num bytes>]
  [--max-write-ops=<num ops>] [--max-write-bytes=<num bytes>]

For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=bucket --bucket=mybucket --max-read-ops=1024 --max-write-bytes=10240

A 0 value for num ops and / or num bytes means that the
specific rate limit attribute check is disabled.

> Get Bucket Rate Limit

Reviewer: please specify the relationship between the use/bukcet ones?
do we block when the first one is reached?

---------------------

Get the current configured rate limit parameters
For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=bucket --bucket=<bucket>

For example:: 

	radosgw-admin ratelimit get --ratelimit-scope=bucket --bucket=mybucket

A 0 value for num ops and / or num bytes means that the
specific rate limit attribute check is disabled.

Enable/Disable Bucket Rate Limit
--------------------------------

Once you set a bucket rate limit, you may enable it. For example:: 

	radosgw-admin ratelimit enable --ratelimit-scope=bucket --bucket=<bucket>

You may disable an enabled bucket rate limit. For example:: 

	radosgw-admin ratelimit disable --ratelimit-scope=bucket --uid=mybucket

Reading / Writing Global Rate Limit Configuration
-------------------------------------------------

> You can read and write global rate limit settings in the period configuration. To

Reviewer: what is the relationship between the global and the per user/bucket limits?
do we block when any of them is reached, or do we override the values with the more specific ones?

Reviewer 2: This doesn't seem to be resolved yet... I don't see an explanation of how global and per user/bucket limits interact, and I agree it's important to explain this.

view the global rate limit settings::

	radosgw-admin global rate limit get

The global rate limit settings can be manipulated with the ``global ratelimit``
counterparts of the ``ratelimit set``, ``ratelimit enable``, and ``ratelimit disable``
commands. ::

	radosgw-admin global ratelimit set --ratelimit-scope bucket --max-read-ops=1024
	radosgw-admin global ratelimit enable --ratelimit-scope bucket

> The global rate limit can configure rate limit scope for all unauthenticated users::

Reviewer: can we also configure them for all users if we use:
```
radosgw-admin global ratelimit enable --ratelimit-scope=user
```

Author: yes

Reviewer: so, can you please add this information/example?

   be restarted for the changes to take effect.

Rate Limit Management
=====================

The Ceph Object Gateway enables you to set rate limits on users and buckets. 
Rate limit includes the maximum number of read ops and write ops per minute
and how many bytes per minute could be written or read per user or per bucket.
Requests that are using GET or HEAD method in the REST request are considered as "read requests", otherwise they are considered as "write requests".
Every Object Gateway tracks per user and bucket metrics separately, these metrics are not shared with other gateways.
That means that the desired limits configured should be divide by the number of active Object Gateways.
For example, if userA should be limited by 10 ops per minute and there are 2 Object Gateways in the cluster,
the limit over userA should be 5 (10 ops per minute / 2 RGWs).
if the requests are ``not`` balanced between RGWs, the rate limit may be underutilized.
For example, if the ops limit is 5 and there are 2 RGWs, ``but`` the Load Balancer send load only to one of those RGWs,
The effective limit would be 5 ops, because this limit is enforced per RGW.

> if there is a limit reached for bucket not for user or vice versa the request would be cancelled as well.

Reviewer: ```suggestion
The request would be rejected if either the user limits or the bucket limits are reached.
```

> It means, if userA has 1 byte read limit per minute and this user tries to GET 1 GB object, the user will be able to do it.

Reviewer: This does not seem related to the previous sentence. Isn't that the definition of the maximum burst you allow?

- **Bucket:** The ``--bucket`` option allows you to specify a rate limit for a
  bucket.

- **User:** The ``--uid`` option allows you to specify a rate limit for a
  user.

- **Maximum Read Ops:** The ``--max-read-ops`` setting allows you to specify
  the maximum number of read ops per minute per RGW. A 0 value disables this setting (which means unlimited access).
  
- **Maximum Read Bytes:** The ``--max-read-bytes`` setting allows you to specify
  the maximum number of read bytes per minute per RGW. A 0 value disables this setting (which means unlimited access).

- **Maximum Write Ops:** The ``--max-write-ops`` setting allows you to specify
  the maximum number of write ops per minute per RGW. A 0 value disables this setting (which means unlimited access).
  
- **Maximum Write Bytes:** The ``--max-write-bytes`` setting allows you to specify
  the maximum number of write bytes per minute per RGW. A 0 value disables this setting (which means unlimited access).
 
- **Rate Limit Scope:** The ``--ratelimit-scope `` option sets the scope for the rate limit.

>   The options are ``bucket`` , ``user`` and ``anonymous``. Bucket rate limit apply to buckets. 

Reviewer 3: the extra space in `--ratelimit-scope ` is causing a rendering issue

Author: I've removed the extra space

  The user rate limit applies to a user. Anonymous applies to an unauthenticated user.
  Anonymous scope is only available for global rate limit.

Set User Rate Limit
-------------------

Before you enable a rate limit, you must first set the rate limit parameters.
For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=user --uid=<uid> <[--max-read-ops=<num ops>] [--max-read-bytes=<num bytes>]
  [--max-write-ops=<num ops>] [--max-write-bytes=<num bytes>]>

For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=user --uid=johndoe --max-read-ops=1024 --max-write-bytes=10240

A 0 value for num ops and / or num bytes means that the
specific rate limit attribute check is disabled.

Get User Rate Limit
-------------------

Get the current configured rate limit parameters
For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=user --uid=<uid>

For example:: 

> 	radosgw-admin ratelimit get --ratelimit-scope=user --uid=johndoe

Reviewer 3: i'm confused about the two separate examples here. the first one says 'ratelimit set' - should we just remove that one?

Author: I've changed the upper example to get

Get Bucket Rate Limit
---------------------

Get the current configured rate limit parameters
For example:: 

	radosgw-admin ratelimit set --ratelimit-scope=bucket --bucket=<bucket>

For example:: 

	radosgw-admin ratelimit get --ratelimit-scope=bucket --bucket=mybucket

A 0 value for num ops and / or num bytes means that the
specific rate limit attribute check is disabled.

Enable/Disable Bucket Rate Limit
--------------------------------

Once you set a bucket rate limit, you may enable it. For example:: 

	radosgw-admin ratelimit enable --ratelimit-scope=bucket --bucket=<bucket>

You may disable an enabled bucket rate limit. For example:: 

	radosgw-admin ratelimit disable --ratelimit-scope=bucket --uid=mybucket

Reading / Writing Global Rate Limit Configuration
-------------------------------------------------

You can read and write global rate limit settings in the period configuration. To
view the global rate limit settings::

	radosgw-admin global rate limit get

The global rate limit settings can be manipulated with the ``global ratelimit``
counterparts of the ``ratelimit set``, ``ratelimit enable``, and ``ratelimit disable``

> commands. ``Per user and per bucket ratelimit configuration is overriding the global configuration``::

Reviewer 3: i don't think this `Per user and per bucket ratelimit` sentence should be highlighted

Author: I've removed the highlight
